### ISSUE TITLE
numpy.place() not implemented 

### ISSUE THREAD

**Issue by Hellisotherpeople** [cupy/cupy#1690]
https://docs.scipy.org/doc/numpy/reference/generated/numpy.place.html

This is the function in question. Seems like it should be easy to implement on the GPU. I'm being forced to rewrite some code that uses it in my attempt to GPUify some numpy code.

**Comment by shreyasbapat** [comment]
Hi! 
Can I work on this one?

**Comment by clemisch** [comment]
Assuming `arr`, `mask` and `vals` have the same shape, this can be implemented easily with `cupy.ElementwiseKernel`.

```python
import cupy as cp

kernel_code = """
if (mask) 
    arr = vals;
"""

kernel = cp.ElementwiseKernel(
    "T vals, bool mask",
    "T arr",
    kernel_code,
    no_return=True
)

arr = cp.ones(20, dtype=float32)
mask = cp.arange(20) % 3 == 0
vals = cp.arange(20, dtype=float32)

kernel(vals, mask, arr)

print(arr)
# [ 0.  1.  1.  3.  1.  1.  6.  1.  1.  9.  1.  1. 12.  1.  1. 15.  1.  1. 18.  1.]
```

If `vals` is of a different shape, you could repeat or slice it in a thin wrapper and then pass it to the `kernel`. 

Or is there something I am missing about `place`?

**Comment by kmaehashi** [comment]
@shreyasbapat Thank you,

**Comment by kmaehashi** [comment]
@shreyasbapat Thank you, could you try implementing it?

**Comment by kmaehashi** [comment]
@clemisch This is not the same behavior as NumPy.

> vals : 1-D sequence
>> Values to put into a. Only the first N elements are used, where N is the number of True values in mask. If vals is smaller than N, it will be repeated, and if elements of a are to be masked, this sequence must be non-empty.

https://docs.scipy.org/doc/numpy/reference/generated/numpy.place.html

**Comment by kmaehashi** [comment]
@clemisch Ah I see, just understood what you mean.
In that case `cupy.where` may also work (although it creates a copy of the array instead of modifying it).

**Comment by clemisch** [comment]
BTW, it has been implemented in a pull request: https://github.com/cupy/cupy/pull/1787
This has not been merched yet to the main `cupy` branch. In the meantime, you could `git checkout` the branch: https://github.com/grafi-tt/cupy/tree/implement-put-and-place

**Comment by shreyasbapat** [comment]
I will surely do that @kmaehashi  if the issue has not been taken by someone else! :)

**Comment by shreyasbapat** [comment]
Is it taken or being worked upon?

**Comment by clemisch** [comment]
As I said in my previous comment:

> BTW, it has been implemented in a pull request: #1787
This has not been merched yet to the main cupy branch. In the meantime, you could git checkout the branch: https://github.com/grafi-tt/cupy/tree/implement-put-and-place

This has already been implemented as passed the test. The `cupy` devs only need to merge the branch. You could of course implement it yourself but I'm not sure whether it's worth your time.

**Comment by shreyasbapat** [comment]
Yeah! True! That's what I was asking for. Thanks @clemisch :)

**Comment by kmaehashi** [comment]
Sorry, I removed `contribution-welcome` as the PR is ongoing.

See https://docs-cupy.chainer.org/en/latest/reference/comparison.html for what we currently have/don't have. Contributions are always welcomed for missing ones.

**Comment by leofang** [comment]
seems this issue can be closed now as #1787 is merged.

**Comment by kmaehashi** [comment]
Closing as it's implemented. Thanks for the heads-up @leofang!
https://docs-cupy.chainer.org/en/latest/reference/generated/cupy.place.html#cupy.place
